[WPE] When a specified WPE_DISPLAY is not available, fail, instead of silently using a fallback

nikolaszimmermann opened the pull request:
#### 3f572031b825b1cb4502e59b719d2e8b125a4509
<pre>
[WPE] When a specified WPE_DISPLAY is not available, fail, instead of silently using a fallback
<a href="https://bugs.webkit.org/show_bug.cgi?id=277165">https://bugs.webkit.org/show_bug.cgi?id=277165</a>

Reviewed by Michael Catanzaro.

If the environemnt variable &apos;WPE_DISPLAY&apos; is explicitely set to a value, e.g. &apos;wpe-display-wayland&apos;
we should fail, if the Wayland connection fails or the platform is unavailable instead of silently
choosing a fallback.

* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_default):

Canonical link: <a href="https://commits.webkit.org/281468@main">https://commits.webkit.org/281468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45e10e96540b608249c68bcc2538f62b2f3941f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60028 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12580 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63946 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10558 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47048 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10766 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48636 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7369 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62059 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/36715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29478 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/33420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9478 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9506 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65678 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3958 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9369 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55990 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51973 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56139 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3285 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8987 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35189 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36271 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37359 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36015 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -186,6 +186,7 @@ WPEDisplay* wpe_display_get_default(void)
                     return;
                 }
                 g_warning("Failed to connect to display of type %s: %s", extensionName, error->message);
+                return nullptr;
```
```suggestion
                return;
```

nikolaszimmermann replied:
Ouch, the "return;" above is problematic, not this one -- `wpe_display_get_default` is supposed to return a WPEDisplay pointer.

The function is used from WebKitWebView:
```
WebKit/UIProcess/API/glib/WebKitWebView.cpp:        priv->display = wpe_display_get_default();
WebKit/UIProcess/API/glib/WebKitWebView.cpp:            priv->display = wpe_display_get_default();
```

nikolaszimmermann replied:
Ouch, forget my comment, misread -- we're inside a lambda.

mcatanzaro reviewed (approved):
This is fine, but the best way to write this would be to use `g_error()` which is the fatal never-returns version of `g_warning()` or `g_critical()`. Then you don't need the `CRASH()` or the `return` at the end.

nikolaszimmermann committed:
[WPE] When a specified WPE_DISPLAY is not available, fail, instead of silently using a fallback
https://bugs.webkit.org/show_bug.cgi?id=277165

Reviewed by Michael Catanzaro.

If the environemnt variable 'WPE_DISPLAY' is explicitely set to a value, e.g. 'wpe-display-wayland'
we should fail, if the Wayland connection fails or the platform is unavailable instead of silently
choosing a fallback.

* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpe_display_get_default):

Canonical link: https://commits.webkit.org/281468@main
